### ISSUE TITLE
Operator Should Re-Create Sts when k8s API blocks Sts updates

### ISSUE THREAD

**Issue by AdheipSingh** [druid-io/druid-operator#76]
- So most common issue is when i need to expand sts volume for Historical ( or even MM ), as of now here's how its done:

1. apply new druid-cr, operator can't re-concile new configs since k8s does not allow mutating anything accept podSpec.
2. So i delete sts using --cascad=false to make sure pod doesnt get deleted and  only sts gets deleted.
3. Operator on next re-conciles creates the new sts with new configs.
4. I manually expand pvc ( sc supports volume expansion )

What i would like to automate here, is the second step. I can easily delete sts using operator with cascade, what i am not sure at what events will the operator know that k8s api will not update the sts. If we can get this, this will be a big help.

@himanshug any thoughts what can help us trigger this command ?

**Comment by AdheipSingh** [comment]
i am wondering if operator can check for this error 
```
Failed to update [StatefulSet:druid-cluster-historical] due to [StatefulSet.apps \"druid-cluster-historical\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden.].
```

- if the operator can do something like this
```
if desiredPVCSize() != currentPVCSize() {
       sdk.delete(sts, deleteOptions{casade})
}
```
OR

- do a linear search in err.Error() string to check for "Forbidden" keyword ( seems a hacky approach, should work though, unless k8s changes the error msg which is unlikely since sts is apps/v1 gvk) This way i ll be able to update anything sts whether labels, pvc size etc.
```
 if err == "Forbidden" {
       sdk.delete(sts, deleteOptions{casade})
}
```

**Comment by himanshug** [comment]
Assuming, It is *always* safe to `delete sts` in pvc size update case. Since there is no generic fix, due to safety reasons, We should only fix the specific use case i.e.

```
if `changed pvc size in spec` != `current pvc size in spec` AND `everything-else-still-the-same` {
  sdk.delete(sts,..)
}
```

also, we can skip the check for string error as that is not part of k8s api contract and can be different for different k8s versions anyways.

we should document that specific thing in the docs . If users want to change pvc size, they should update only the pvc size before apply any other updates.

**Comment by AdheipSingh** [comment]
yeah its very safe to delete sts ONLY WITH ```--cascade=false``` .

Why do we need to evalute ```everything-else-still-the-same``` ? just confused on that, for specific pvc case i guess we need to just evalute ``currentPvcSize and ChangedPvcSize``` .

**Comment by himanshug** [comment]
> Why do we need to evalute everything-else-still-the-same ? just confused on that, for specific pvc case i guess we need to just evalute currentPvcSize and ChangedPvcSize .

Intention is to have the path of automatic delete-and-recreate-sts just to support pvc size update and nothing else as we are not sure about safety of the operation. if user updates pvc size ***and*** some other fields then we shouldn't follow the path introduced here... we can add more things for that path on a case-by-case basis consciously. 
my assumption is that when k8s doesn't allow updating sts for certain field change then that is a special situation needing user attention, we can make exceptions to those as we encounter them.

**Comment by AdheipSingh** [comment]
got it , makes sense. 
Thanks for the explanation.

Ill try to make this work. :)
